server/src/main/java/org/opensearch/indices/tiering/HotToWarmTieringService.java

```diff
@@ -0,0 +1,875 @@
+        }
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        // TODO add handling for master switch, dangling indices, master reload
+        if (event.routingTableChanged()) {
+            if (!indexTieringInfoMap.isEmpty()) {
+                updateIndexShardStatus(event.state());
+            }
+            if (!requestUuidToRequestContext.isEmpty()) {
+                completeRequestLevelTiering();
+            }
+        }
+    }
+
+    /**
+     * Checks and completes the request level tiering for
+     * the request uuids in requestUuidToRequestContext map.
+     */
+    void completeRequestLevelTiering() {
+        for (String requestUuid : requestUuidToRequestContext.keySet()) {
+            final TieringRequestContext requestContext = requestUuidToRequestContext.get(requestUuid);
+            if (requestContext.isRequestProcessingComplete()) {
+                logger.info("Tiering is completed for the request [{}]", requestContext);
+                for (Index index : indexTieringInfoMap.keySet()) {
+                    indexTieringInfoMap.remove(index);
+                }
+                requestUuidToRequestContext.remove(requestUuid);
+                if (requestContext.getRequest().waitForCompletion()) {
+                    requestContext.getListener().onResponse(requestContext.constructHotToWarmTieringResponse());
+                }
+            }
+        }
+    }
+
+    /**
+     * Checks if the given node is a search node.
+     * @param node the node to check.
+     * @return true if the node is a search node, false otherwise.
+     */
+    boolean isSearchNode(final DiscoveryNode node) {
+        return node != null && node.isSearchNode();
+    }
+
+    /**
+     * Updates the shard tiering status based on the shard state and the node that it is assigned to.
+     * @param shard the shard to update the status for.
+     * @param clusterState the cluster state.
+     * @return the updated shard tiering status.
+     */
+    ShardTieringStatus updateShardTieringStatus(final ShardRouting shard, final ClusterState clusterState) {
+        final boolean isShardFoundOnSearchNode = isSearchNode(clusterState.getNodes().get(shard.currentNodeId()));
+        final boolean isShardRelocatingToSearchNode = isSearchNode(clusterState.getNodes().get(shard.relocatingNodeId()));
+        ShardTieringState tieringState;
+        String reason = null;
+        if (shard.started() && isShardFoundOnSearchNode) {
+            tieringState = ShardTieringState.SUCCESSFUL;
+        } else if (shard.unassigned()) {
+            tieringState = ShardTieringState.INIT;
+            reason = "Shard is unassigned due to " + shard.unassignedInfo().getReason();
+        } else if (((shard.initializing() || shard.started()) && !isShardFoundOnSearchNode)
+            || (shard.relocating() && !isShardRelocatingToSearchNode)) {
+                tieringState = ShardTieringState.FAILED;
```

Comment:
Along with the shard allocation retries, I think we should also add some tiering retry logic at service level as well.

Reply:
Can you share the issue link here as well for reference?

Reply:
here u go: https://github.com/opensearch-project/OpenSearch/issues/14982